### PR TITLE
feat(API): Add endpoint to find metrics of a service

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR intends to add an endpoint to find metrics of a service.

`URI: /monitoring/hosts/{hostId}/services/{serviceId}/metrics`

```
API Contract:

type: array
items:
  type: object
  properties:
    id:
      type: int
    name:
      type: string
    unit:
      type: string
    current_value:
      type: float
      nullable:true
    warning_high_threshold:
      type: float
      nullable: true
    warning_low_threshold:
      type: float
      nullable: true
    critical_high_threshold:
      type: float
      nullable: true
    critical_low_threshold:
      type: float
      nullable: true
```
**Fixes** # MON-24211

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [x] 24.04.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/Metric/Application/UseCase/FindMetricsByService/FindMetricsByService.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                $accessGroups = $this->accessGroupRepository->findByContact($this->user);
+                $metrics = $this->metricRepository->findByHostIdAndServiceIdAndAccessGroups(
+                    $request->hostId,
+                    $request->serviceId,
+                    $accessGroups,
+                    $this->requestParameters
+                );
+            }[] === $metrics
+                ? $presenter->presentResponse(new NotFoundResponse('metrics'))
+                : $presenter->presentResponse($this->createResponse($metrics));
```
Comment (arnapou) [inline centreon/src/Core/Metric/Application/UseCase/FindMetricsByService/FindMetricsByService.php:71]: 1. Indentation problem
2. hidden if is not a nice idea : please explicit a if or use a match

File: centreon/src/Core/Metric/Infrastructure/Repository/DbMetricFactory.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+declare(strict_types=1);
+
+namespace Core\Metric\Infrastructure\Repository;
+
+use Core\Metric\Domain\Model\Metric;
+
+class DbMetricFactory
+{
+    public static function createFromRecord(array $record): Metric
```
Comment (arnapou) [inline centreon/src/Core/Metric/Infrastructure/Repository/DbMetricFactory.php:30]: Missing phpstan array phpdoc.

File: centreon/src/Core/Metric/Infrastructure/Repository/DbReadMetricRepository.php [hunk 3]
```diff
@@ -135,6 +160,59 @@ public function findServicesByMetricNamesAndAccessGroupsAndRequestParameters(
...
+     *  }
+     * > $records
+     *
+     * @return Metric[]
+     */
+    private function createMetricsFromRecords(array $records): array
+    {
+        if ([] === $records) {
+            return [];
+        }
```
Comment (arnapou) [inline centreon/src/Core/Metric/Infrastructure/Repository/DbReadMetricRepository.php:206]: useless `if`.

Comment (jeremyjaouen) [reply centreon/src/Core/Metric/Infrastructure/Repository/DbReadMetricRepository.php:206]: early return

**Review by dmyios — DISMISSED** [review]

**Review by adr-mo — APPROVED** [review]

**Review by adr-mo — DISMISSED** [review]

**Review by adr-mo — DISMISSED** [review]

**Review by dmyios — APPROVED** [review]

**Review by adr-mo — APPROVED** [review]

File: centreon/src/Core/Metric/Infrastructure/API/FindMetricsByService/FindMetricsByServiceController.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+     * @param int $serviceId
+     * @param FindMetricsByService $useCase
+     * @param FindMetricsByServicePresenter $presenter
+     */
+    public function __invoke(
+        int $hostId,
+        int $serviceId,
+        FindMetricsByService $useCase,
+        FindMetricsByServicePresenter $presenter
+    ) {
```
Comment (arnapou) [inline centreon/src/Core/Metric/Infrastructure/API/FindMetricsByService/FindMetricsByServiceController.php:42]: Missing function's return type declaration

File: centreon/src/Core/Metric/Infrastructure/API/FindMetricsByService/FindMetricsByServicePresenter.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+use Core\Metric\Application\UseCase\FindMetricsByService\MetricDto;
+
+final class FindMetricsByServicePresenter extends AbstractPresenter implements FindMetricsByServicePresenterInterface
+{
+    public function __construct(protected PresenterFormatterInterface $presenterFormatter)
+    {
+        parent::__construct($presenterFormatter);
+    }
+
+    public function presentResponse(FindMetricsByServiceResponse|ResponseStatusInterface $response): void
```
Comment (arnapou) [inline centreon/src/Core/Metric/Infrastructure/API/FindMetricsByService/FindMetricsByServicePresenter.php:40]: Parameter's name changed from '$data' to '$response'
:point_right: The interface has not the same name.

File: centreon/src/Core/Metric/Infrastructure/Repository/DbMetricFactory.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+namespace Core\Metric\Infrastructure\Repository;
+
+use Core\Metric\Domain\Model\Metric;
+
+class DbMetricFactory
+{
+    /**
+     * @param array{
+     *    id: int,
+     *    name: string
```
Comment (arnapou) [inline centreon/src/Core/Metric/Infrastructure/Repository/DbMetricFactory.php:33]: Missing comma
```suggestion
     *    name: string,
```

**Review by arnapou — APPROVED** [review]

**Review by adr-mo — APPROVED** [review]

**Review by dmyios — APPROVED** [review]

**Pull request merged**